src/metabase/api/field.clj

```diff
@@ -71,24 +72,70 @@
   [id]
   (let [field (api/read-check Field id)]
     (if-not (field-should-have-field-values? field)
-      {:values {} :human_readable_values {}}
-      (create-field-values-if-needed! field))))
+      {:values []}
+      {:values (create-field-values-if-needed! field)})))
 
+(api/defendpoint POST "/:id/dimension"
+  "If `Field`'s special type derives from `type/Category`, or its base type is `type/Boolean`, return
+   all distinct values of the field, and a map of human-readable values defined by the user."
+  [id :as {{dimension-type :type dimension-name :name human_readable_field_id :human_readable_field_id} :body}]
+  {dimension-type         (s/enum "internal" "external")
+   dimension-name         su/NonBlankString
+   human_readable_field_id (s/maybe s/Int)}
+  (let [field (api/read-check Field id)]
+    (if-let [dimension (Dimensions :field_id id)]
+      (db/update! Dimensions (:id dimension)
+        {:type dimension-type
+         :name dimension-name
+         :human_readable_field_id human_readable_field_id})
+      (db/insert! Dimensions
+                  {:field_id id
+                   :type dimension-type
+                   :name dimension-name
+                   :human_readable_field_id human_readable_field_id}))
+    (Dimensions :field_id id)))
+
+(defn validate-human-readable-pairs
+  "Human readable values are optional, but if present they must be
+  present for each field value. Throws if invalid, returns a boolean
+  indicating whether human readable values were found."
+  [value-pairs]
+  (let [human-readable-missing? #(= ::not-found (get % 1 ::not-found))
+        has-human-readable-values? (not-any? human-readable-missing? value-pairs)]
+    (api/check (or has-human-readable-values?
+                   (every? human-readable-missing? value-pairs))
+      [400 "If remapped values are specified, they must be specified for all field values"])
+    has-human-readable-values?))
 
-;; TODO - not sure this is used anymore
-(api/defendpoint POST "/:id/value_map_update"
+(defn- update-field-values [field-value-id value-pairs]
+  (let [human-readable-values? (validate-human-readable-pairs value-pairs)]
+    (api/check-500 (db/update! FieldValues field-value-id
+                     :values (map first value-pairs)
+                     :human_readable_values (when human-readable-values?
+                                              (map second value-pairs))))))
+
+(defn- create-field-values
```

Comment:
Since function has side effects I would vote it end in a bang, e.g. `create-field-values!`. Calling it twice would not be the same as calling it once; in other words, it wouldn't be safe inside an STM transaction. https://github.com/bbatsov/clojure-style-guide#changing-state-fns-with-exclamation-mark